feature(unlock-app) - referrer fees

easyrun42 opened the pull request:
https://github.com/unlock-protocol/unlock/assets/136458308/464cd14f-327d-4ff9-aa45-85fe41585d01

easyrun42 commented:
@julien51 As requested the UI will look like this:

- We removed the toggle next to the fee itself
- Address part will be disabled until the toggle is turned on

easyrun42 commented on packages/unlock-js/src/web3Service.ts:
```diff
@@ -899,26 +899,6 @@ export default class Web3Service extends UnlockService {
-  /**
-   * Returns referrer fees
-   */
-  async referrerFees({
-    lockAddress,
-    network,
-    address,
-  }: {
-    lockAddress: string
-    network: number
-    address: string
-  }) {
-    const lockContract = await this.getLockContract(
-      lockAddress,
-      this.providerForNetwork(network)
-    )
-    const referrerFees = await lockContract.referrerFees(address)
-    return ethers.BigNumber.from(referrerFees).toNumber()
-  }
```
No longer needed since we don't use this anymore. This calls the referrerFees from smart contract, but in our case we use subgraph 👍

searchableguy replied:
I would keep this since someone may be depending on this downstream. The library is published and used by others.

julien51 replied:
+1

easyrun42 replied:
Good call 👍. Adding it back 🙏

easyrun42 commented on unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:
```diff
@@ -16,108 +24,170 @@ interface UpdateReferralFeeProps {
+            <LoadingIcon />
+          </div>
+        ) : referralFees.length ? (
+          <div className="w-full grid gap-5 grid-cols-1 mt-3">
+            {referralFees.map(({ id, referrer, fee }) => {
+              const feeNumber = Number(fee)
+
+              return (
+                <div
+                  key={id}
+                  className="w-full overflow-x-auto flex justify-between items-center border border-gray-300 p-3 rounded-xl"
+                >
+                  <h2 className="text-md font-medium break-words">
+                    {referrer}
+                  </h2>
+                  <h2 className="text-md pl-3 font-medium break-words">
+                    ({feeNumber / 100}%)
+                  </h2>
+                </div>
+              )
+            })}
+          </div>
```
I wanted to put this in `@unlock-protocol/ui` but i guess it should be added if it's a major component used throughout the UI ? Let me know if this is fine or if i should put this in `@unlock-protocol/ui` 👍

searchableguy replied:
The way we add components in the UI library is if something is repeated at least a couple times in the main codebase. 

I think it's fine to not worry about it for now.

easyrun42 replied:
Sounds good🙏

searchableguy commented on unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:
```diff
@@ -16,108 +24,170 @@ interface UpdateReferralFeeProps {
-      address: ZERO,
+      address: fields?.referralAddress,
       feeBasisPoint: fields?.referralFeePercentage * 100,
     })
   }
 
-  const getReferrerFees = async () => {
-    return await web3Service.referrerFees({
-      lockAddress,
-      network,
-      address: ZERO,
-    })
+  const getLock = async () => {
+    const service = new SubgraphService()
+    return service.lock(
+      {
+        where: {
+          id: lockAddress,
+        },
+      },
+      {
+        network,
+      }
+    )
   }
```
May wanna move this out of react component or wrap it in useCallback?

searchableguy commented on packages/unlock-js/src/subgraph/schema.graphql:
```diff
@@ -25,6 +26,11 @@ query allLocks(
     maxKeysPerAddress
     createdAtBlock
     totalKeys
+    referrerFees(skip: $referrerFeeSkip) {
+      id
+      referrer
+      fee
+    }
   }
 }
```
bump up the version in package.json of unlock-js and add it changelog.

julien51 reviewed (commented):
Can you please show what happens when the referrer address is *not* set (a fee should apply to _any_ address).
That's the current flow in the app and we need to make sure it still works!

easyrun42 commented on unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:
```diff
@@ -16,108 +24,170 @@ interface UpdateReferralFeeProps {
+            setIsReferralAddressEnabled(enabled)
+          }}
+        />
+
```
Right. This is supposed to be a boolean whenever we toggle the referrer address field. I renamed this to `isReferralAddressToggled` so it would make more sense ✅

easyrun42 commented on unlock-app/src/hooks/useReferrerFee.ts:
```diff
@@ -0,0 +1,63 @@
+    })
+  }
+
+  const setReferrerFee = useMutation(setReferrerFeeForLock)
+
+  const isSettingReferrerFee = setReferrerFee.isLoading
+
+  const getLock = async () => {
+    const service = new SubgraphService()
+    return service.lock(
+      {
+        where: {
+          id: lockAddress,
+        },
+      },
+      {
+        network,
+      }
+    )
+  }
+
+  const {
+    isLoading,
+    data: referralFeesData,
+    error,
+  } = useQuery(
+    ['getLock', lockAddress, network, setReferrerFee.isSuccess],
+    async () => getLock()
+  )
+
+  const data = referralFeesData?.referrerFees || []
+
+  return {
+    isLoading,
+    data,
+    setReferrerFee,
+    isSettingReferrerFee,
+    error,
+  }
+}
```
Here is the `useReferrerFee` which will live outside of the `<UpdateReferralFee />`. This way it will be easier to get mutations and query data ✨

easyrun42 commented on packages/unlock-js/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # Changes
 
+# 0.39.3
+
+- add `referrerFees` to `lock`.
+
```
Bumped the version to `0.39.3` and updated the version in `package.json`

easyrun42 commented on unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:
```diff
@@ -16,108 +22,157 @@ interface UpdateReferralFeeProps {
+          <div className="w-full grid gap-5 grid-cols-1 mt-3">
+            {referralFees.map(({ id, referrer, fee }) => {
+              const feeNumber = Number(fee)
+              const isAddressLinkedToAnyReferrer = referrer === zeroAddress
+
+              return (
+                <div
+                  key={id}
+                  className="w-full overflow-x-auto flex justify-between items-center border border-gray-300 p-3 rounded-xl"
+                >
+                  <h2 className="text-md font-medium break-words">
+                    {isAddressLinkedToAnyReferrer
+                      ? 'Referral fee applied to any referrer'
+                      : referrer}
+                  </h2>
+                  <h2 className="text-md pl-3 font-medium break-words">
+                    ({feeNumber / 100}%)
+                  </h2>
+                </div>
```
This one is saying basically is the address `0x0000000000000000000000000000000000000000`. This means it applies to any referrer.

easyrun42 commented on unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:
```diff
@@ -16,108 +22,157 @@ interface UpdateReferralFeeProps {
 
   const onSubmit = async (fields: FormProps) => {
-    const setReferrerFeePromise = setReferrerFeeMutation.mutateAsync(fields)
+    const setReferrerFeePromise = setReferrerFee.mutateAsync({
+      ...fields,
+      referralAddress: referralAddress || zeroAddress,
+    })
```
This will apply to the address you've entered or apply the address to "any" referrer. A.k.a `0x0000000000000000000000000000000000000000`

easyrun42 commented:
@searchableguy Here are the updates:
- Bumped the `unlock-js` version to `0.39.3` and added it changelog
- Moved out the referrer fee mutations and queries to a new hook called `useReferrerFee` 👍

@julien51 Here are the updates from your side:
- If doesn't set an address the referrer fee will be set to any referrer
- If user puts an address then the referrer will be applied to that address

Here is a quick demo:

https://github.com/unlock-protocol/unlock/assets/136458308/40e94014-90d5-437f-95bf-a246d4ac2adf

blahkheart commented:
@easyrun42 Can you please provide your discord username and eth address so you can be nominated for the DAO retroactive funding for your contributions

julien51 commented on packages/unlock-js/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+# 0.39.3
+
+- add `referrerFees` to `lock`.
```
```suggestion
- add `referrerFees` to `lock` in subgraphs
```

julien51 commented:
@easyrun42 How are we looking here? I am worried this is dragging a bit :)

KosmosKey commented on unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:
```diff
@@ -16,108 +23,181 @@ interface UpdateReferralFeeProps {
+                  className="justify-self-left"
+                  width="20"
+                  height="20"
+                  alt="Apple Wallet"
+                  src={`/images/illustrations/refresh.svg`}
+                />
+              }
+            >
+              Refresh
+            </Button>
+          ) : null}
```
In case the subgraph is not full in sync we can add a "Refresh" button which will refresh the referrer fees list. This can improve overall the UX experience 😁

julien51 commented:
I think the UI can be improved a bit because it is still quite unclear. 
By default, let's just keep the toggle but hide the field for a custom address if the toggle is disabled. Let's also rename the toggle to "Set a referrer for a custom address". Like this:

<img width="680" alt="Screenshot 2023-09-20 at 11 16 37 AM" src="https://github.com/unlock-protocol/unlock/assets/17735/f906d460-d543-4a2a-beeb-b565c4b6bae2">

Now when the toggle is enabled:

<img width="686" alt="Screenshot 2023-09-20 at 11 17 20 AM" src="https://github.com/unlock-protocol/unlock/assets/17735/82fb5725-4eb8-4a29-8038-661c608e74f5">

And make sure there is a label _above_ the referrer field that says "Referrer fee (%)".

Thanks!

easyrun42 commented:
> I think the UI can be improved a bit because it is still quite unclear. By default, let's just keep the toggle but hide the field for a custom address if the toggle is disabled. Let's also rename the toggle to "Set a referrer for a custom address". Like this:
> 
> <img alt="Screenshot 2023-09-20 at 11 16 37 AM" width="680" src="https://user-images.githubusercontent.com/17735/269333743-f906d460-d543-4a2a-beeb-b565c4b6bae2.png">
> Now when the toggle is enabled:
> 
> <img alt="Screenshot 2023-09-20 at 11 17 20 AM" width="686" src="https://user-images.githubusercontent.com/17735/269334048-82fb5725-4eb8-4a29-8038-661c608e74f5.png">
> And make sure there is a label _above_ the referrer field that says "Referrer fee (%)".
> 
> Thanks!


@julien51  Here is the new version 👍

https://github.com/unlock-protocol/unlock/assets/136458308/8f9f70d2-7079-4a28-9adf-3a54c40e8e50

julien51 reviewed (approved):
LGTM!
